[USH-3701] N-per-row-repeat for repeat group

akashkj opened the pull request:
## Product Description
Adds a new appearance attribute `n-per-row-repeat` to the `Group` element

## Technical Summary
https://dimagi-dev.atlassian.net/browse/USH-3701

A new appearance attribute is supported for app builder - `n-per-row-repeat` which can be applied to a `Group` such that any `Repeat` placed inside this `Group` will have the property styling `n-per-row` applied automatically.

As is, appearance attributes configured on `Repeat` are not passed in Formplayer's response. Because of this, limitation, we are using an appearance attribute applied to a `Group` to define nested Repeat appearance. There is [ongoing work](https://dimagi.atlassian.net/browse/USH-4077?jql=text%20~%20%22repeat%20appearance%20attribute*%22) to support appearance attribute direclty on Repeat.

Below are examples of form configuration and the resulting form layout.
CONFIGURATION 1:
![Screen Shot 2024-02-15 at 5 10 55 PM](https://github.com/dimagi/commcare-hq/assets/88759246/0b36e531-1b01-438a-a45a-3f4ef86e36df)
![Screen Shot 2024-02-15 at 5 09 59 PM](https://github.com/dimagi/commcare-hq/assets/88759246/017ba063-54d4-48f7-8722-b81fcd469780)

CONFIGURATION 2:
![Screen Shot 2024-02-15 at 5 12 12 PM](https://github.com/dimagi/commcare-hq/assets/88759246/bc9c4715-4ee6-47d5-9118-027a3e3f52ef)
![Screen Shot 2024-02-15 at 5 09 32 PM](https://github.com/dimagi/commcare-hq/assets/88759246/840a6484-369d-4d6d-8457-56e56e6d0c53)

## Feature Flag
NA

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
Tested locally with combinations of `Repeat` and `Group`.
### Automated test coverage

Added test that `Repeat` are grouped correctly when nested within a `n-per-row-repeat` Group.

### QA Plan
https://dimagi.atlassian.net/browse/QA-6032?focusedCommentId=313072&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-313072

### Migrations
NA

<!-- Please link to any past code changes that are coordinated with this migration -->

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

Commit:
Repeat group n-per-row-repeat attr

Commit:
Add n-per-row style to repeat Group

Commit:
Merge remote-tracking branch 'origin/akj/ush-3855-group-per-row-attr' into akj/ush-3701-n-per-row-repeat-group

Commit:
Add test to apply styling to inner group when outer group has n-per-row-repeat styling

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -137,6 +137,21 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+    function isStyleNPerRowRepeat(style) {
+        var pattern = /^\d+-per-row-repeat$/;
+        return pattern.test(style);
+    }
```
could we add `per-row-repeats` as a constant similar we do for other appearance attributes? 

Could this actually reuse most of what's defined for `calculateElementWidth `?

akashkj replied:
https://github.com/dimagi/commcare-hq/pull/33979/commits/5b6b1fe56b4bb555babdcadcf0bb168de3d27361

Jtang-1 replied:
Is there a way to also combine this into `calculateElementWidth`?

akashkj replied:
Isn't `calculateElementWidth` providing a different return i.e. width of each item while this only returns a boolean if style is present?

akashkj replied:
Cam also do something like get the element width back and check if its less than 12. If it is that means it is having that style applied and can use this condition to derive a boolean expression. Does this sound plausible?

Jtang-1 replied:
oh I see, my bad. I guess there isn't a good way to combine it without making it convoluted. Then just a nit: is this similar enough to how `getMatchingStyles` is used?

akashkj replied:
It is already using `getMatcingStyles`

Probably can simplify like: `return getMatchingStyles(pattern, styleStr).length > 0`

Jtang-1 replied:
Ya basically the same as what's done for ` Question.prototype.stylesContains `

Jtang-1 replied:
https://github.com/dimagi/commcare-hq/pull/33979/commits/caa335077b3e082cc0e96b8f226d1fc84a483821

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -313,6 +328,13 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
         }
 
         for (let child of json.children) {
+            if (isNPerRowRepeatElement(child)) {
```
Could all of this be incorporated as part of the `if (child.type === constants.GROUP_TYPE) ` condition?

akashkj replied:
Do you mean to say put this condition within https://github.com/dimagi/commcare-hq/pull/33979/files/cdbbdeba26cd4e11b9b47677863cf2e6515ac94d#diff-fa78ad0c884f5acb33ba17fb2142ba438de13b9bb043c65b1687111cfca48190R345 ?

Jtang-1 replied:
yes, I think it'll have to be changed a bit but it should be able to be combined

akashkj replied:
Can you specify in little more detail what is intended to be done here? Shouldn't adding style `n-per-row` be done to the repeat group's `Group` before reaching the condition `if (child.type === constants.GROUP_TYPE)` ?

Jtang-1 replied:
you're adding `n-per-row` to the children of the element defined as `n-per-row-repeat` right? I think when we get to `if (child.type === constants.GROUP_TYPE)`, we can also check if it also contains the `n-per-row-repeat` appearance attribute. If so, then we populate its children with the respective `n-per-row`

akashkj replied:
https://github.com/dimagi/commcare-hq/pull/33979/commits/ba593a503571ab1a54b2f59c30982a4d56127423

Jtang-1 replied:
ya that's what i was imaging - does that work with testing?

akashkj replied:
Yes, it is working.

Commit:
Refactor function to check matching pattern for n-per-row-repeat

Commit:
Merge remote-tracking branch 'origin/akj/ush-3855-group-per-row-attr' into akj/ush-3701-n-per-row-repeat-group

Jtang-1 committed:
Merge branch 'akj/ush-3855-group-per-row-attr' into akj/ush-3701-n-per-row-repeat-group

esoergel commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -137,6 +137,22 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+    }
+
+    function isNPerRowRepeatElement(element) {
+        return element.type === constants.GROUP_TYPE && element.hasOwnProperty('style') && element.style.hasOwnProperty('raw') && isStyleNPerRowRepeat(element.style.raw)
```
FYI looks like this is erroring on staging:
![image](https://github.com/dimagi/commcare-hq/assets/2367539/9509693d-8931-4030-a1b7-b2f658d7cbb5)
I suspect its that `element.style` exists, but is `null`.

akashkj replied:
Thanks @esoergel for highlighting this. In local testing I had not come across this scenario where style exists but is null. Anyhow, I have pushed the fix for it and issue seems to be resolved.

Commit:
Modify stlye check condition

Commit:
Merge remote-tracking branch 'origin/master' into akj/ush-3701-n-per-row-repeat-group

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -137,6 +137,31 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+        return integerPart + '-per-row';
+    }
+
+    function isNPerRowRepeatElement(element) {
```
This seems like a code smell to me. Can't this be done in a way similar to how we're checking for any other appearance attribute without doing all these specific checks?

akashkj replied:
Do you mean something like:
```
function isNPerRowRepeatElement(element) {
    const styleStr = (style) ? ko.utils.unwrapObservable(element.style.raw) : null;
    if(styleStr) {
        return isStyleNPerRowRepeat(styleStr)
    } else {
        return false
    }
}
```

Jtang-1 replied:
ya something like that, and can probably just pass in the `style` instead of the whole `element` right?

akashkj replied:
Not moving content of `inisStyleNPerRowRepeat` as it is used in `getNPerRowStyleFromRepeatStyle` as well.

Modified in https://github.com/dimagi/commcare-hq/pull/33979/commits/3306c7c5d73ebe0c42d0d40c03213b5fe785945c

Jtang-1 replied:
https://github.com/dimagi/commcare-hq/pull/33979/commits/caa335077b3e082cc0e96b8f226d1fc84a483821

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -137,6 +137,31 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+    }
+
+    function getNPerRowStyleFromRepeatStyle(style) {
+        var parts = style.split('-');
```
would there would an issue if style includes multiple space separate appearance attributes?

akashkj replied:
https://github.com/dimagi/commcare-hq/pull/33979/commits/101adeb05550bf60b9fc237abdaccb3fbbd615dc

Commit:
Move check for repeat element inside existing group check

Commit:
Refactor

Commit:
Get n-per-row-repeat style from a group of styles

Jtang-1 commented:
An additional general note, I saw you added a test which is great. I think it would also be beneficial to test the end result grouping, similar to what you did [here](https://github.com/dimagi/commcare-hq/blob/101adeb05550bf60b9fc237abdaccb3fbbd615dc/corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js#L117)

Commit:
Refactor isNPerRowRepeatElement

Commit:
Modify reepat group test to add more assertions

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -321,6 +344,13 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
                 if (child.type === constants.GROUP_TYPE) {
+                    if (isNPerRowRepeatElement(child.style)) {
+                        for (let groupChild of child.children) {
+                            if (groupChild.type === constants.GROUP_TYPE) {
```
Given what we learned [here](https://dimagi.atlassian.net/browse/QA-6043) about the type assigned to repeat groups in the formplayer response, I feel like this logic needs to change to handle that.

Jtang-1 replied:
applied grouping for `repeat-juncture` type elements https://github.com/dimagi/commcare-hq/pull/33979/commits/cd09d773b099ba55e0f762b59e2aa2a22ea17ae3

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/spec/form_ui_spec.js:
```diff
@@ -169,6 +169,39 @@ hqDefine("cloudcare/js/form_entry/spec/form_ui_spec", function () {
+        it('Should add n-per-row style to inner Group when using n-per-row-repeat and render Repeat Group and Question grouped by row', function () {
+            let styleObj = {raw: '3-per-row-repeat'};
+
+            let g0 = fixtures.groupJSON({
```
Since this is about grouping repeat groups, shouldn't this also insert `repeatJSON` as its children to test that it is correclty grouped?

And I forget, is the intention for `3-per-row-repeat` to also group subgroups? Like if the form builder is 
- Group1 (3-per-row-repeat)
       -Group2
       - Repeat Group1 
       - Repeat Group2
  
"Group2" isn't supposed to be bundled with "Repeat Group1" and "Repeat Group2" right?

Jtang-1 replied:
Leaving a note: no it's not supposed to also group the `Group` nested within `n-per-row-repeat`. It should apply to only `Repeat` questions.

Jtang-1 committed:
bug: n-per-row style shouldn't overwrite existing styles

style also does not have a 'raw' key if no style is configured

Jtang-1 committed:
bug: should only group sub-group types if it is a repeatable

In form configuration, when a repeat-group is placed in a group, the
repeat-group's definition in the json is assigned type 'sub-group' with
'repeatble' key set to "true". The same occurs if 'repeat-count' is
configured

Jtang-1 committed:
bug: should only group sub-group types if it is a repeatable

In form configuration, when a repeat-group is placed in a group, the
repeat-group's definition in the json is assigned type 'sub-group' with
'repeatble' key set to "true". The same occurs if 'repeat-count' is
configured

Jtang-1 committed:
refactor: define pattern as constant for '-per-row' appearance attributes

Jtang-1 committed:
prefactor: pulls out logic that's not specific to Question

Jtang-1 committed:
refactor: uses more generic functions for matching styles

Jtang-1 committed:
refactor: calculate value once instead of every loop

Jtang-1 committed:
feat: applies grouping to json representation of form elements

Jtang-1 committed:
feat: adds column width class for repeat-juncture types

Jtang-1 committed:
style: indentation

Jtang-1 committed:
refactor: both types sub-group and repeat-juncture need to recursively call and group their children

Jtang-1 committed:
clarify comment

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -321,6 +344,13 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+                    if (isNPerRowRepeatElement(child.style)) {
+                        for (let groupChild of child.children) {
+                            if (groupChild.type === constants.GROUP_TYPE) {
+                                groupChild.style.raw = getNPerRowStyleFromRepeatStyle(child.style.raw);
```
[d634eb4](https://github.com/dimagi/commcare-hq/pull/33979/commits/d634eb416fa0b2c40b02c40a11d81d2b120fd293)

Jtang-1 committed:
test: correct tests to account for additional nesting

Repeat questions are now nested within a Group-Element-Tile-Row.
So an additional 'children' call needs to be made to account for
additional nesting

Jtang-1 committed:
bug: need to account for additional nesting caused from grouped-element-tile-row

Jtang-1 committed:
bug: check 'required' exists before calling. addresses failing tests

Jtang-1 committed:
test: updates test to check that Repeat are grouped if they are nested in a Group configured with n-per-row-repeat

n-oer-row-repeat should not affect nested Group. It should affect only nested Repeat

Jtang-1 committed:
update comment

Jtang-1 commented:
I was reviewing this, but then I picked it up from where Akash left off so I added others as reviewers. FYI It's ready for review.

AddisonDunn reviewed (commented):
I'm mostly caught up on the logic where the styling is applied.

Jtang-1 committed:
refactor: adds the style directly to parent's child instead of the grandparent's grandchild

Jtang-1 committed:
refactor: extract logic of applying n-per-row style to children of n-oer-row-repeat style group

Jtang-1 committed:
refactor: use general function

Jtang-1 committed:
style: improve readability and return empty string for consistency in return type

Jtang-1 committed:
test: ensures existing styles are not overwritten

Jtang-1 committed:
style: create constant since it corresponds to a specific "style"

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -313,21 +331,36 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+                            }
+                        }
+                    }
+                }
```
Ah good point [f72b95b](https://github.com/dimagi/commcare-hq/pull/33979/commits/f72b95b1f93fe1591c1675c586308a11c27ecd16) and [62af412](https://github.com/dimagi/commcare-hq/pull/33979/commits/62af412cc1728957cf238a13cbc421f30fcbaf1f)

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -95,6 +95,15 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+
+    function stylesContains(pattern, style) {
+        return stylesContaining(pattern, style).length > 0;
+    }
```
[3d167d4](https://github.com/dimagi/commcare-hq/pull/33979/commits/3d167d41d1021c3c49e73fc12ff51640867183fe)

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -137,6 +146,15 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+            perRowRepeatStyle = matchingPerRowRepeatStyles.length === 0 ? null : matchingPerRowRepeatStyles[0];
+        if (perRowRepeatStyle) {
+            const integerPart = perRowRepeatStyle.split('-')[0];
+            return integerPart + '-per-row';
```
[c6bc514](https://github.com/dimagi/commcare-hq/pull/33979/commits/c6bc5143c3a4ce0bebebf5172bec72b54e71fecc)

AddisonDunn reviewed (approved):
For the record I think this PR could benefit from some QA focused on repeat groups (with and without the appearance attribute)

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -155,6 +155,23 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+                }
+            }
+        }
+    }
```
I don't think it should be in `Container`. Don't have a great reasoning other than that it's not specific to `Container` object.

Jtang-1 commented on corehq/apps/cloudcare/static/cloudcare/js/form_entry/form_ui.js:
```diff
@@ -137,6 +146,34 @@ hqDefine("cloudcare/js/form_entry/form_ui", function () {
+                // Detects configured repeat groups within the form. If a repeat group has a 'repeat-count' configured,
+                // the Formplayer response designates the key 'type' as 'sub-group' and 'repeatable' as 'true'.
+                if ((groupChild.type === constants.GROUP_TYPE && groupChild.repeatable === "true") || groupChild.type === constants.REPEAT_TYPE) {
+                    if (_.has(groupChild, 'style') && groupChild.style && groupChild.style.raw) {
```
that would work, but i'm just going to keep it as is since I think the checks are pretty straightforward.

Jtang-1 commented:
Thanks for the review Addison! There was also a QA check specifically for repeat groups. I just added the ticket to the PR description
